### PR TITLE
Add new command to update security group name

### PR DESCRIPTION
**Pull request by ravening**
## Description

By default, once we create a security group we cant change its name.
In this feature, we introduce a new API command "updateSecurityGroup"
which allows us to rename the security group name. Although we can't
change the name of the "default" security group.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

Navigate to "Network" section in the main window and select "Security groups" from the drop-down menu in the top left section

Click on "Add Security group" to create a new security group
Screenshot 2019-12-03 at 16 04 49

Once the security group is created you can rename it using the edit button

Screenshot 2019-12-03 at 16 06 33

## How Has This Been Tested?

1. Navigate to "Network" tab in the main UI.
2. Select "Security groups" from the drop-down menu in the top left.
3. Click on "Add Security group" on the top right to create a new security group.
4. Give any suitable name and create it.
5. Now select the security group you just created and click on edit button to enter the new name
Screenshot 2019-12-03 at 16 02 45
6. This will rename the security group to the new name
7. Now again try to edit the security group name and enter blank spaces. You will get error message which indicates that security group name cannot be empty.
Screenshot 2019-12-03 at 16 04 01

## Testing

New integration test case has been added which validates different test cases in updating the security group name. This can be run using

```nosetests --with-marvin --marvin-config=<config file> test/integration/component/test_update_security_group.py```

### CODE REVIEW

File: server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java [hunk 1]
```diff
@@ -858,6 +859,10 @@ public SecurityGroupVO createSecurityGroup(CreateSecurityGroupCmd cmd) throws Pe
         Account caller = CallContext.current().getCallingAccount();
         Account owner = _accountMgr.finalizeOwner(caller, cmd.getAccountName(), cmd.getDomainId(), cmd.getProjectId());
 
+        if (name.trim().equals("")) {
```
Comment (wido) [inline server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java]: Maybe use the commons check ifNotEmpty? We use this on multiple locations in the CloudStack code

Comment (ravening) [reply server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java]: Updated the code

**Comment by rhtyd** [comment]
@ravening few points:
- Please follow and implement based more modern API implementation such as https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/admin/acl/UpdateRoleCmd.java#L34 (put version, use APINAME etc.)
- Move the test to smoke instead of component if it's a fast enough test
- Add the new marvin test to `.travis.yml` (sort alphabetically)

**Comment by ravening** [comment]
> @ravening few points:
> 
> * Please follow and implement based more modern API implementation such as https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/admin/acl/UpdateRoleCmd.java#L34 (put version, use APINAME etc.)
> * Move the test to smoke instead of component if it's a fast enough test
> * Add the new marvin test to `.travis.yml` (sort alphabetically)

@rhtyd I have made the necessary changes. Please review it again

**Comment by rhtyd** [comment]
thanks @ravening 
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
I did not test it.

File: api/src/main/java/org/apache/cloudstack/api/command/user/securitygroup/UpdateSecurityGroupCmd.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        return name;
+    }
+
+    /////////////////////////////////////////////////////
+    /////////////// API Implementation///////////////////
+    /////////////////////////////////////////////////////
+
+    @Override
+    public String getCommandName() {
+        return s_name;
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/user/securitygroup/UpdateSecurityGroupCmd.java]: I'd rather see it constructed form the API name .toLower()+"response". no biggy

File: api/src/main/java/org/apache/cloudstack/api/command/user/securitygroup/UpdateSecurityGroupCmd.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public long getEntityOwnerId() {
+        SecurityGroup securityGroup = _entityMgr.findById(SecurityGroup.class, getId());
+        if (securityGroup != null) {
+            return securityGroup.getAccountId();
+        }
+
+        return Account.ACCOUNT_ID_SYSTEM; // no account info given, parent this command to SYSTEM so ERROR events are tracked
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/user/securitygroup/UpdateSecurityGroupCmd.java]: isn't this an exceptional situation? wouldn't there always be at least root admin responsible for creating a sec-group?

File: server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java [hunk 4]
```diff
@@ -1117,6 +1122,60 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
...
+
+        // check permissions
+        _accountMgr.checkAccess(caller, null, true, group);
+
+        return Transaction.execute(new TransactionCallback<SecurityGroupVO>() {
+            @Override
+            public SecurityGroupVO doInTransaction(TransactionStatus status) {
+                SecurityGroupVO group = _securityGroupDao.lockRow(groupId, true);
+                if (group == null) {
+                    throw new InvalidParameterValueException("Unable to find security group by id " + groupId);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/security/SecurityGroupManagerImpl.java]: the message here could better be "unable to acquire lock for security group with id " + groupId

**Comment by DaanHoogland** [comment]
tested and works, just one remark. I think this is a generic details page problem in cloudstack but after saving, the old name is still used instead of the altered one.

**Comment by andrijapanicsb** [comment]
> tested and works, just one remark. I think this is a generic details page problem in cloudstack but after saving, the old name is still used instead of the altered one.

log out and in, or force refresh page - I recall similar issues on other parts of the gui

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
will be regression tested in Trillian, when possible. thx

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@ravening can you enable the specific UI action in primate: https://github.com/apache/cloudstack-primate/blob/master/src/config/section/network.js#L188

**Pull request merged**